azkaban-common/src/main/java/azkaban/scheduler/MissedSchedulesManager.java

```diff
@@ -0,0 +1,216 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package azkaban.scheduler;
+
+import azkaban.Constants;
+import azkaban.project.Project;
+import azkaban.project.ProjectManager;
+import azkaban.trigger.builtin.ExecuteFlowAction;
+import azkaban.utils.DaemonThreadFactory;
+import azkaban.utils.Emailer;
+import azkaban.utils.Props;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import java.text.MessageFormat;
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Miss Schedule Manager is a failure recover manager for schedules. Schedule might be missed to execute due to
+ * Trigger Scanner thread busy or web server down. This manager maintains a configurable fixed sized thread pool
+ * to execute tasks including send email notification or back execute flow.
+ * */
+@Singleton
+public class MissedSchedulesManager {
+  private static final Logger LOG = LoggerFactory.getLogger(MissedSchedulesManager.class);
+
+  private ExecutorService missedSchedulesExecutor;
+  private final boolean missedSchedulesManagerEnabled;
+  private final ProjectManager projectManager;
+  // TODO: Refactored to AlerterHolder to support missed scheduler use case
+  private final Emailer emailer;
+
+  // parameters to adjust MissScheduler
+  private final int threadPoolSize;
+  private final int DEFAULT_THREAD_POOL_SIZE = 5;
+
+  @Inject
+  public MissedSchedulesManager(final Props azkProps,
+      final ProjectManager projectManager,
+      final Emailer emailer) {
+    this.projectManager = projectManager;
+    this.emailer = emailer;
+    this.threadPoolSize = azkProps.getInt(Constants.MISSED_SCHEDULE_THREAD_POOL_SIZE, DEFAULT_THREAD_POOL_SIZE);
+    this.missedSchedulesManagerEnabled = azkProps.getBoolean(Constants.MISSED_SCHEDULE_MANAGER_ENABLED, false);
+    if (this.missedSchedulesManagerEnabled && threadPoolSize <= 0) {
+      String errorMsg = "MissedSchedulesManager is enabled but thread pool size is not positive: " + threadPoolSize;
+      LOG.error(errorMsg);
+      throw new RuntimeException(errorMsg);
+    }
+  }
+
+  /**
+   * Put timestamps for Missed executions along with email notifications into Task Queue.
+   *
+   * @param missedScheduleTimesInMs, timestamps of missed schedule
+   * @param action, execute flow action
+   * @param backExecutionEnabled, a schedule config from user, the default is false
+   * */
+  public boolean addMissedSchedule(final List<Long> missedScheduleTimesInMs,
```

Comment:
Why is the return type a boolean, shouldn't it be throwing exceptions in case of any errors anyway?

Reply:
I was planning to encapsulate all the failure handling logic inside this manager class, without leaking to other components. Since this service is robust, it also reduces the chances of introducing regression for existing system. The callers of this manager only know whether they successfully put the notification tasks in queue or not.